C/wrc20.c

```diff
@@ -0,0 +1,133 @@
+
+
+// to avoid includes from libc, just hard-code things
+typedef unsigned char uint8_t;
+typedef int uint32_t;
+typedef unsigned long long uint64_t;
+
+
+// types used for Ethereum stuff
+typedef uint8_t* bytes; // an array of bytes with unrestricted length
+typedef uint8_t bytes32[32]; // an array of 32 bytes
+typedef uint8_t address[20]; // an array of 20 bytes
+typedef unsigned __int128 u128; // a 128 bit number, represented as a 16 bytes long little endian unsigned integer in memory
+//typedef uint256_t u256; // a 256 bit number, represented as a 32 bytes long little endian unsigned integer in memory
+typedef uint32_t i32; // same as i32 in WebAssembly
+typedef uint32_t i32ptr; // same as i32 in WebAssembly, but treated as a pointer to a WebAssembly memory offset
+typedef uint64_t i64; // same as i64 in WebAssembly
+
+// functions for ethereum stuff
+void useGas(i64 amount);
+void getCaller(i32ptr* resultOffset);
+   // memory offset to load the address into (address)
+i32 getCallDataSize();
+void callDataCopy(i32ptr* resultOffset, i32 dataOffset, i32 length);
+   // memory offset to load data into (bytes), the offset in the input data, the length of data to copy
+void revert(i32ptr* dataOffset, i32 dataLength);
+void finish(i32ptr* dataOffset, i32 dataLength);
+void storageStore(i32ptr* pathOffset, i32ptr* resultOffset);
+void storageLoad(i32ptr* pathOffset, i32ptr* resultOffset);
+   //the memory offset to load the path from (bytes32), the memory offset to store/load the result at (bytes32)
+void printMemHex(i32ptr* offset, i32 length);
+void printStorageHex(i32ptr* key);
+
+
+i64 reverse_bytes(i64 a){
+  i64 b = 0;
+  b += (a & 0xff00000000000000)>>56;
```

Comment:
Replace `+=` with `|=` and magic happens: https://godbolt.org/z/GYL1L0
Unfortunately, wasm does not have bswap support.

Reply:
Cool!

But Wasm has no bswap opcode. I would have to look at the output wasm, but don't think there is a difference since addition + and or | both use the ALU in one cycle.